From 8ee5adda1ac8c78b59fe5790da5f0510b067a079 Mon Sep 17 00:00:00 2001
From: everyx <lunt.luo@gmail.com>
Date: Mon, 1 Sep 2025 11:56:09 +0800
Subject: [PATCH 2/3] messageTray: Add timeout watchdog for notification hiding
 animation

The animation in `_hideNotification()` can occasionally get stuck, failing to call its `onComplete` handler. This leaves the notification in the HIDING state indefinitely.

Add a `GLib.timeout_add()` as a watchdog. If the animation doesn't complete within a reasonable time after it starts, the timeout forces the state to HIDDEN and completes the cleanup.

The timeout is removed upon successful animation completion.

Closes: https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/6006
---
 js/ui/messageTray.js | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/js/ui/messageTray.js b/js/ui/messageTray.js
index 30502bfe1..bae529893 100644
--- a/js/ui/messageTray.js
+++ b/js/ui/messageTray.js
@@ -1241,6 +1241,14 @@ export const MessageTray = GObject.registerClass({
 
         const duration = animate ? ANIMATION_TIME : 0;
         this._notificationState = State.HIDING;
+
+        const timeoutId = GLib.timeout_add(GLib.PRIORITY_DEFAULT,
+            duration + 1000, () => {
+                log('Animation timeout, forcing notification cleanup');
+                this._forceHideNotificationCompleted();
+                return GLib.SOURCE_REMOVE;
+            });
+
         this._bannerBin.ease({
             opacity: 0,
             duration,
@@ -1251,6 +1259,7 @@ export const MessageTray = GObject.registerClass({
             duration,
             mode: Clutter.AnimationMode.EASE_OUT_BACK,
             onComplete: () => {
+                GLib.source_remove(timeoutId);
                 this._notificationState = State.HIDDEN;
                 this._hideNotificationCompleted();
                 this._updateState();
@@ -1258,6 +1267,14 @@ export const MessageTray = GObject.registerClass({
         });
     }
 
+    _forceHideNotificationCompleted() {
+        if (this._notificationState !== State.HIDDEN) {
+            this._notificationState = State.HIDDEN;
+            this._hideNotificationCompleted();
+            this._updateState();
+        }
+    }
+
     _hideNotificationCompleted() {
         let notification = this._notification;
         this._notification = null;
-- 
2.51.0

